File: models/customized-services/CustomizedServiceRequest.ts
import mongoose, { Schema, Document } from "mongoose";

export interface ICustomizedServiceRequest extends Document {
  customId: string; // CSR0001, CSR0002, ...
  roomName: string;
  residentEmail: string;
  title: string;
  description?: string;
  status: "new" | "accepted" | "completed" | "no-show" | "canceled";
  assignee?: {
    name: string;
    staffId: string;
    profilePic?: string;
  };
  createdAt: Date;
  updatedAt: Date;
}

const customizedServiceRequestSchema = new Schema<ICustomizedServiceRequest>(
  {
    customId: { type: String, unique: true },
    roomName: { type: String, required: true },
    residentEmail: { type: String, required: true },
    title: { type: String, required: true },
    description: { type: String, trim: true },

    status: {
      type: String,
      enum: ["new", "accepted", "completed", "no-show", "canceled"],
      default: "new",
    },

    assignee: {
      name: String,
      staffId: String,
      profilePic: String,
    },
  },
  { timestamps: true }
);

// 🔢 Auto-generate customId like CSR0001, CSR0002, ...
customizedServiceRequestSchema.pre("save", async function (next) {
  if (this.isNew && !this.customId) {
    try {
      const model = mongoose.model<ICustomizedServiceRequest>(
        "CustomizedServiceRequest"
      );
      const lastDoc = await model.findOne().sort({ createdAt: -1 });

      let nextNumber = 1;
      if (lastDoc && lastDoc.customId) {
        const match = lastDoc.customId.match(/\d+$/);
        if (match) nextNumber = parseInt(match[0], 10) + 1;
      }

      // Decide padding based on number size (e.g., CSR0001 → CSR00001 after 9999)
      const padding = nextNumber > 9999 ? 5 : 4;
      this.customId = `CSR${String(nextNumber).padStart(padding, "0")}`;
    } catch (error) {
      return next(error as Error);
    }
  }
  next();
});

// Avoid model overwrite error in dev
if (
  process.env.NODE_ENV !== "production" &&
  mongoose.models.CustomizedServiceRequest
) {
  delete mongoose.models.CustomizedServiceRequest;
}

export const CustomizedServiceRequest = mongoose.model<ICustomizedServiceRequest>(
  "CustomizedServiceRequest",
  customizedServiceRequestSchema
);

export default CustomizedServiceRequest;

File: luxeli_app/lib/features/specials/models/customized_service_request.dart
class CustomizedServiceRequest {
  final String id;
  final String customId;
  final String roomName;
  final String residentEmail;
  final String title;
  final String? description;
  final String status;
  final Assignee? assignee;
  final DateTime createdAt;
  final DateTime updatedAt;

  CustomizedServiceRequest({
    required this.id,
    required this.customId,
    required this.roomName,
    required this.residentEmail,
    required this.title,
    this.description,
    required this.status,
    this.assignee,
    required this.createdAt,
    required this.updatedAt,
  });

  factory CustomizedServiceRequest.fromJson(Map<String, dynamic> json) {
    return CustomizedServiceRequest(
      id: json['_id'] as String? ?? json['id'] as String,
      customId: json['customId'] as String? ?? '',
      roomName: json['roomName'] as String? ?? '',
      residentEmail: json['residentEmail'] as String? ?? '',
      title: json['title'] as String? ?? '',
      description: json['description'] as String?,
      status: json['status'] as String? ?? 'new',
      assignee: json['assignee'] != null
          ? Assignee.fromJson(json['assignee'] as Map<String, dynamic>)
          : null,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'customId': customId,
      'roomName': roomName,
      'residentEmail': residentEmail,
      'title': title,
      'description': description,
      'status': status,
      'assignee': assignee?.toJson(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class Assignee {
  final String name;
  final String staffId;
  final String? profilePic;

  Assignee({required this.name, required this.staffId, this.profilePic});

  factory Assignee.fromJson(Map<String, dynamic> json) {
    return Assignee(
      name: json['name'] as String? ?? '',
      staffId: json['staffId'] as String? ?? '',
      profilePic: json['profilePic'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {'name': name, 'staffId': staffId, 'profilePic': profilePic};
  }
}
